client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1001 @@
+        KAA_LOG_ERROR(self->logger, KAA_ERR_NOT_FOUND, "Topic with id = %d not found.", *topic_id);
+        return KAA_ERR_NOT_FOUND;
+    }
+    *topic = kaa_list_get_data(topic_node);
+    return KAA_ERR_NONE;
+}
+
+static bool kaa_find_topic_state_by_id(void *topic_state, void *id)
+{
+    kaa_topic_state_t *state = (kaa_topic_state_t *)topic_state;
+    return state->topic_id == *(uint64_t *)id;
+}
+
+kaa_error_t kaa_calculate_topic_listener_id(kaa_topic_listener_t *listener, uint32_t *listener_id)
+{
+    KAA_RETURN_IF_NIL2(listener, listener_id, KAA_ERR_BADPARAM);
+
+    const uint32_t prime = 31;
+
+    *listener_id = 1;
+    *listener_id = prime * (*listener_id) + (ptrdiff_t)listener->context;
+    *listener_id = prime * (*listener_id) + (ptrdiff_t)listener->callback;
+     return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_calculate_notification_listener_id(kaa_notification_listener_t *listener, uint32_t *listener_id)
+{
+    KAA_RETURN_IF_NIL2(listener, listener_id, KAA_ERR_BADPARAM);
+
+    const uint32_t prime = 31;
+
+    *listener_id = 1;
+    *listener_id = prime * (*listener_id) + (ptrdiff_t)listener->context;
+    *listener_id = prime * (*listener_id) + (ptrdiff_t)listener->callback;
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_notification_manager_get_size(kaa_notification_manager_t *self, size_t *expected_size)
+{
+    KAA_RETURN_IF_NIL2(self, expected_size, KAA_ERR_BADPARAM);
+    *expected_size = KAA_EXTENSION_HEADER_SIZE;
+    *expected_size += sizeof(uint32_t); //state sqn size
+
+    if (self->topic_states) {
+        *expected_size += sizeof(uint32_t); //field id + reserved + count
+        *expected_size += (sizeof(uint64_t) + sizeof(uint32_t)) * kaa_list_get_size(self->topic_states);
+    }
+
+    if (self->uids) {
+        *expected_size += sizeof(uint32_t); //field id + reserved + count
+        kaa_list_t *uid_node = self->uids;
+        size_t uids_size = sizeof(uint32_t) * kaa_list_get_size(self->uids);
+        while (uid_node) {
+            kaa_notifications_uid_t *uid = (kaa_notifications_uid_t *)kaa_list_get_data(uid_node);
+            uids_size += kaa_aligned_size_get(uid->length);
+            uid_node = kaa_list_next(uid_node);
+        }
+        *expected_size += uids_size;
+    }
+
+    if (self->subscriptions) {
+        *expected_size += sizeof(uint32_t); //meta data for subscribe commands
+        *expected_size += kaa_list_get_size(self->subscriptions) * sizeof(uint64_t);
+    }
+    if (self->unsubscriptions) {
+        *expected_size += sizeof(uint32_t); //meta data for unsubscribe commands
+        *expected_size += kaa_list_get_size(self->unsubscriptions) * sizeof(uint64_t);
+    }
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_notification_manager_request_serialize(kaa_notification_manager_t *self, kaa_platform_message_writer_t *writer)
+{
+    KAA_RETURN_IF_NIL2(self, writer, KAA_ERR_BADPARAM);
+
+    size_t size = 0;
+    kaa_error_t err = kaa_notification_manager_get_size(self, &size);
```

Comment:
Calling twice kaa_notification_manager_get_size() is  an unreasonable overload. It's better to cache the size.